[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations 

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations
* #86407 [11/N] Update all_to_all with CPU/CUDA implementations
* #86368 [10/N] Update barrier with CPU/CUDA implementations
* #86166 [9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations
* **#84423 [8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations**
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations

### Changes
- Updates for the allgather collective

### Context
https://github.com/pytorch/pytorch/issues/86225

H-Huang committed:
[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations 

[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -106,8 +88,7 @@ TORCH_LIBRARY(c10d, m) {
   m.def(
-      "allgather_",
-      dispatch(c10::DispatchKey::CompositeExplicitAutograd, allgather_));
+      "allgather_(Tensor[][] output_tensors, Tensor[] input_tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, int timeout) -> (Tensor[][], __torch__.torch.classes.c10d.Work)");
```
Hey @bdhirsh, this is a convoluted question so bear with me.

I am having an issue with implementing the dispatch definition for [all_gather](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather). I cannot specify the 2nd argument as `Tensor[] input_tensors` because it throws an error. Example script:
```python
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import os

os.environ["MASTER_ADDR"] = "localhost"
os.environ["MASTER_PORT"] = "29500"
os.environ["TORCH_SHOW_CPP_STACKTRACES"] = "1"

def work(rank, world_size):
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    print(f"finished creating process group for {rank}")
    t = torch.zeros(2, device="cuda")
    t_list = [torch.zeros(2, device="cuda") for _ in range(1)]
    print(f"{t_list=}")
    dist.all_gather(t_list, t) # ERRORS HERE

if __name__ == "__main__":
    world_size = 2
    mp.spawn(
        work,
        nprocs=world_size,
        args=(world_size,),
    )
```
throws error
```
NotImplementedError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors), but no fallback function is registered for schema c10d::allgather_.  This usually means that this function requires a non-empty list of Tensors, or that you (the operator writer) forgot to register a fallback function.  Available functions are [CPU, CUDA, BackendSelect, Python, Functionalize, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradMPS, AutogradXPU, AutogradHPU, AutogradLazy, Tracer, AutocastCPU, AutocastCUDA, Batched, VmapMode, PythonTLSSnapshot].

CPU: registered at ../torch/csrc/distributed/c10d/OpsImpl.cpp:274 [kernel]
CUDA: registered at ../torch/csrc/distributed/c10d/OpsImpl.cpp:278 [kernel]
BackendSelect: fallthrough registered at ../aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
...
```

I noticed that in documentation [`Tensor[]` translates to type of ArrayRef<Tensor>](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#:~:text=Tensor%5B%5D.%20A%20Tensor%5B%5D%20argument%20translates%20into%20a%20C%2B%2B%20argument%20of%20type%20ArrayRef%3CTensor%3E%20(a.k.a.%20TensorList)). When I turn change our allgather implementations to fit that signature there is no more error. For example,
```
# from
const std::vector<at::Tensor>& input_tensors
# to
at::TensorList input_tensors
```
However, it would be ideal not to change our implementation definitions and we may suffer performance costs. I also noticed that in  docs it mentions the return type [`Tensor[]` can be translated to std::vector<Tensor>](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#:~:text=The%20following%20are%20permissible%20on%20ReturnType,types%20Tensor%20and%20std%3A%3Avector%3CTensor%3E) so I am curious why `std::vector<Tensor>` is not supported.

Two questions:
1. I am curious why that particular error is thrown, why does it say `There were no tensor arguments to this function`
2. Can `Tensor[] -> std::vector<Tensor>` be supported in the dispatch arguments?

bdhirsh replied:
Hey!

> I am curious why that particular error is thrown, why does it say There were no tensor arguments to this function

Agreed that this seems like a pretty poor error message for getting the schema slightly wrong. Can you file an issue for it (with a quick minimal repro)?

 > However, it would be ideal not to change our implementation definitions and we may suffer performance costs. I also noticed that in docs it mentions the return type [Tensor[] can be translated to std::vector](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#:~:text=The%20following%20are%20permissible%20on%20ReturnType,types%20Tensor%20and%20std%3A%3Avector%3CTensor%3E) so I am curious why std::vector<Tensor> is not supported.

So, the idea with `Tensor[]` getting represented in C++ as `ArrayRef<Tensor>` (aka a `TensorList`) is actually specifically for performance. `ArrayRef` is non-owning, which makes passing it around from function to function cheaper (it's also faster than passing a `std::vector<Tensor>` by reference, since that gives you two layers of indirection).

Is the problem that you're eventually calling a distributed API that expects a vector<Tensor>, so you're forced to pay for the conversion from `ArrayRef<Tensor>` -> `vector<Tensor>`? (which is probably small but non-zero).

H-Huang replied:
> Agreed that this seems like a pretty poor error message for getting the schema slightly wrong. Can you file an issue for it (with a quick minimal repro)?

I'm not sure how to reproduce this issue on current operators. So here is a tiny PR that you can check out and run to reproduce the error: https://github.com/pytorch/pytorch/pull/85115

> Is the problem that you're eventually calling a distributed API that expects a vector, so you're forced to pay for the conversion from ArrayRef<Tensor> -> vector<Tensor>? (which is probably small but non-zero).

Yes, exactly! I thought the `ArrayRef<Tensor>` -> `vector<Tensor>`  conversion may be expensive, especially if the tensor is large?

> So, the idea with Tensor[] getting represented in C++ as ArrayRef<Tensor> (aka a TensorList) is actually specifically for performance. ArrayRef is non-owning, which makes passing it around from function to function cheaper (it's also faster than passing a std::vector<Tensor> by reference, since that gives you two layers of indirection).

I see. That is good to know. Ideally we should change the c10d collectives to take in `TensorList` then, but that may be BC-breaking to some users who've built extensions on these collectives and require some more changes to our backends. I will follow up on this with my team.

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

kwen2501 reviewed (approved):
LGTM! Thanks!

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "




[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40044549).

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "


### Changes
- Updates for the allgather collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40044549](https://our.internmc.facebook.com/intern/diff/D40044549)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40044549).

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "


### Changes
- Updates for the allgather collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40044549](https://our.internmc.facebook.com/intern/diff/D40044549)

[ghstack-poisoned]

H-Huang committed:
Update on "[8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations "


### Changes
- Updates for the allgather collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40044549](https://our.internmc.facebook.com/intern/diff/D40044549)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40044549).

H-Huang commented:
@pytorchbot merge -f "Ran ciflow/trunk and all tests pass, using force to skip meta internal-only check since the files are the same as internal"

H-Huang commented:
@pytorchbot merge -f "Ran ciflow/trunk and all tests pass, using force to skip meta internal-only check since the files are the same as internal"
